lightning/src/ln/features.rs

```diff
@@ -0,0 +1,256 @@
+//! Lightning exposes sets of supported operations through "feature flags". This module includes
+//! types to store those feature flags and query for specific flags.
+
+use std::{cmp, fmt};
+use std::result::Result;
+use std::marker::PhantomData;
+
+use ln::msgs::DecodeError;
+use util::ser::{Readable, Writeable, Writer};
+
+mod sealed { // You should just use the type aliases instead.
+	pub struct InitContext {}
+	pub struct NodeContext {}
+	pub struct ChannelContext {}
+
+	/// An internal trait capturing the various feature context types
+	pub trait Context {}
+	impl Context for InitContext {}
+	impl Context for NodeContext {}
+	impl Context for ChannelContext {}
+
+	pub trait DataLossProtect: Context {}
+	impl DataLossProtect for InitContext {}
+	impl DataLossProtect for NodeContext {}
+
+	pub trait InitialRoutingSync: Context {}
+	impl InitialRoutingSync for InitContext {}
+
+	pub trait UpfrontShutdownScript: Context {}
+	impl UpfrontShutdownScript for InitContext {}
+	impl UpfrontShutdownScript for NodeContext {}
+}
+
+/// Tracks the set of features which a node implements, templated by the context in which it
+/// appears.
+pub struct Features<T: sealed::Context> {
+	/// Note that, for convinience, flags is LITTLE endian (despite being big-endian on the wire)
+	flags: Vec<u8>,
+	mark: PhantomData<T>,
+}
+
+impl<T: sealed::Context> Clone for Features<T> {
+	fn clone(&self) -> Self {
+		Self {
+			flags: self.flags.clone(),
+			mark: PhantomData,
+		}
+	}
+}
+impl<T: sealed::Context> PartialEq for Features<T> {
+	fn eq(&self, o: &Self) -> bool {
+		self.flags.eq(&o.flags)
+	}
+}
+impl<T: sealed::Context> fmt::Debug for Features<T> {
+	fn fmt(&self, fmt: &mut fmt::Formatter) -> Result<(), fmt::Error> {
+		self.flags.fmt(fmt)
+	}
+}
+
+/// A feature message as it appears in an init message
+pub type InitFeatures = Features<sealed::InitContext>;
+/// A feature message as it appears in a node_announcement message
+pub type NodeFeatures = Features<sealed::NodeContext>;
+/// A feature message as it appears in a channel_announcement message
+pub type ChannelFeatures = Features<sealed::ChannelContext>;
+
+impl InitFeatures {
+	/// Create a Features with the features we support
+	#[cfg(not(feature = "fuzztarget"))]
+	pub(crate) fn our_features() -> InitFeatures {
```

Comment:
Hmm. default is strange IMO because the "our" refers to "rust-lightning's set of supported". Agree "our" is maybe less clear, do you have another idea for a better way to say "the set of features supported by rust-lightning, which should be advertised to our peers".

Reply:
Hmm, given its just pub(crate), I dont think we need to worry too much about overriding some. new() is a bit too generic, too, cause its not clear that its intended specifically for the "this is what i want to send a peer" use-case. Maybe features_we_implement?

Reply:
The suggestion of `default` was mainly to demonstrate this as a convention for specifying typical options. I see how `new` can be too generic especially given we have the more specific `empty`.

Using pronouns like we/our/my in identifiers is bit ambiguous for conveying information in my opinion. Don't get me going on identifiers like `MyFoo`. ;)

I would suggest either `known` or `supported`. That is, these features are either known or supported by the implementation. Using `known` also contrasts with uses of `unknown_bits` in the code. Optionally, adding either `_bits` or `_features` to the end may offer more clarity, though it would be inconsistent with `empty` unless also updated.

Reply:
Gonna go with supported(), seems to capture what we want. Note that it is not guaranteed that all non-unknown fetures are supported(), and is in fact already the case that we "know" of some features which we do not support.